keep camilladsp in background, fix mpd malloc() crash

paolosabatino opened the pull request:
Some rework done to avoid reloading camilladsp on start/stop to keep it working in background as much as possible.
When the FIFO has no more data, the camilladsp process stalls and wait for new data. As soon as new data arrives, the process immediately restart processing it

This fixes the crashing issue with mpd and greatly reduces the gap between tracks, but also introduces a glitch when there is a sampling rate change: `pushState` message may not be sent soon enough by volumio state machine, hence the next track may start with a wrong sampling rate that gets fixed withing a second from the start.

CamillaDSP process handling is in a separate .js file which is imported at runtime to keep things separated from the main FusionDSP code.

balbuze commented:
hello,
I did some tests...
It doesn't work as expected.
sr.log was used to remove filter<samplerate/2 or camilla fails (with some low sample rate track/radio). The file doesn't exists but still called in index.js.
Spotify can't work after a 44.1 file
Spotify can't work with convolution filter and autoswitch samplerate because  a 320KHz instead of 44.1kHzfile is seen (mix between sample rate and bitrate) or maybe due to last Spotify PR?
Number of Input channel must be 2 or it fails with mono (1 channel) track/radio
Have you tested it with a recent mpd version and Qobuz? 
Sometimes skipping track results in a sort of white noise or played at a uncorrect speed..
thanks for your work ;-)

paolosabatino commented:
> sr.log was used to remove filter<samplerate/2 or camilla fails (with some low sample rate track/radio). The file doesn't exists but still called in index.js.

sr.log could not be called because it has been removed from index.js code. I can't see any reference to it [in the source code](https://github.com/volumio/volumio-plugins-sources/blob/48bed6fdcbb4f1da353fb76d8a6523461ee6e1f3/fusion/index.js) the check against half sample rate is still there, just the file has been renamed to better reflect the other informations carrying in.

>  Spotify can't work after a 44.1 file

Didn't test spotify, but local vorbis files at 44.1 and 96khz run perfectly

> Spotify can't work with convolution filter and autoswitch samplerate because a 320KHz instead of 44.1kHzfile is seen (mix between sample rate and bitrate) or maybe due to last Spotify PR?

Perhaps some invalid sample rate comes from pushState event? `320` looks more like the max bitrate of mp3 rather than a sampling rate; in such case, perhaps the spotify plugin needs to be fixed?

> Number of Input channel must be 2 or it fails with mono (1 channel) track/radio
ALSA configuration uses `plug` plugin for the `volumioDsp` device, it should convert anything to stereo and signed 32 bit samples. It also looks quite curious to me that mono did work before, as long as the original configuration [hardcoded 2 channels](https://github.com/volumio/volumio-plugins-sources/blob/a2af3c20a498aac121ce7648ae63580f64fd9dc5/fusion/asound/volumioDsp.postDsp.10.conf#L10) in the camilladsp process spawning

>  Have you tested it with a recent mpd version and Qobuz?
That was my primary testing bed. mpd is updated to 0.23.5 and Qobuz playlists were tested since the crash only happens with such kind of source material (ie: does never happen with local files).

>  Sometimes skipping track results in a sort of white noise or played at a uncorrect speed..
Could you be more specific? I never had white noise in any case skipping tracks; about the uncorrect speed, it is the glitch I refer in the description above, and usually it lasts for less than a second until a `pushState` event arrives or until, for some reason, the camilladsp configuration is reloaded when two adjacent tracks have different sampling rate.

volumio commented:
To detect sampling frequency change, instead of relying on the push state, it would be good to tail or check for changes in the /proc/asound/cardX/pcmxp/ ... 
As soon as there is a change it will be notified there

balbuze commented:
Volumio hook does it as well

paolosabatino commented:
The old code was using the volumio hook to detect the sample rate change; I preserved that but also added, in the `pushState` listener, the detection of samplerate property. 

If a valid samplerate is coming from `pushState` event, it has priority over the volumio hook because (sometimes, not always) `pushState` is contemporary to play event and camilladsp can be reloaded promptly.

Doing tail on `/proc/asound/cardX/...` has to be tested for readiness, I bet it is better and faster than `pushState` workaround if a listener is immediately invoked when sample rate changes.

paolosabatino commented:
update:

I implemented the `tail -f` idea, not on the pseudofile from ALSA but on `/tmp/fusiondsp_stream_params.log` regular file and changed some bits to get rid of `pushState` event at all; now, when there is a samplerate change, camilladsp process is stopped, configuration is updated, then camilladsp process is restarted. When there is no samplerate change, the process service is not stopped at all.

At the moment the sample rate change is much faster than before, and after a quick check against qobuz I still get no crashes from mpd.

The sample rate glitch now can be considered gone; there is another glitch though: very seldom `tail -f` does not detect the changes in the file and thus does not update the sample rate. I still can't find a way to steadily reproduce the issue.

this command line:
`volumio livelog | grep -i 'samplerate' | grep read` 

will show a couple of lines when the samplerate change is detected

paolosabatino commented:
Update:

I managed to refine the sample rate check; now it uses `fs.watch` in place of external `tail` command and added a guard against nearby events that may trigger a race condition while updating camilladsp configuration. 
It should be much more robust now and I didn't experience any missing change in sample rate triggering various local tracks, qobuz tracks and webradios with different sample rates.

Please test and report, thanks!

balbuze commented:
hello!
Thanks for this new works! 
Some short tests, I can't do more today...
Why when a 22050Hz file is played, it is seen as 44.1kHz by camilladsp. Why?
![image](https://github.com/volumio/volumio-plugins-sources/assets/12850139/92531f6f-c55b-4e4e-808f-3cc9a57daf5a)

Will test deeply next couple of days

paolosabatino commented:
>  Why when a 22050Hz file is played, it is seen as 44.1kHz by camilladsp. Why?

Honestly I don't know, you can do `cat /tmp/fusiondsp_stream_params.log` to see what samplerate has been passed to the ALSA subsystem.
In case an invalid samplerate is passed (which means 0 or NaN), a default samplerate of 44.1khz is used.

I didn't test 22khz samples though, so can't be sure what is wrong.

Darmur commented:
@balbuze on what device are you resting the 22050kHz playback?

if it's Integro, it could be due to the changes in the kernel we did for solving the playback issue for FS lower than 44.1khz
The I2S driver accepts only FS between 44.1kHz and 192kHz, if they are higher or lower they are resampled

balbuze commented:
Yes it's integro! I know sample rate<44100 can't be played by integro. But I was not aware something changed in kernel side.

Darmur commented:
this can well explain the 44.1kHz you see in camilladsp

the proof of the pudding would be to try on RPi

paolosabatino commented:
Hmm, I'm having another issue that is related to the MCU on Integro.
After turning the unit off, either by idle timeout or by pushing the button, after some seconds (less than a minute) the unit is turned on again.

`volumio livelog` reports this:

```
Nov 08 21:48:08 integro volumio[5391]: info: MCU Signalled Playback Active 
Nov 08 21:48:08 integro volumio[5391]: info: MCU Signalled Sleep Mode Disabled 
Nov 08 21:48:08 integro sudo[7692]:  volumio : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/usr/bin/xset s off -dpms 
Nov 08 21:48:08 integro sudo[7692]: pam_unix(sudo:session): session opened for user root by (uid=0) 
Nov 08 21:48:08 integro kernel: fb: osd[0] enable: 1 (Xorg) 
Nov 08 21:48:08 integro kernel: fb: osd[0] enable: 1 (Xorg) 
Nov 08 21:48:08 integro sudo[7692]: pam_unix(sudo:session): session closed for user root 
Nov 08 21:48:08 integro volumio[5391]: info: CoreStateMachine::pushState 
Nov 08 21:48:08 integro volumio[5391]: info: CorePlayQueue::getTrack 1 
Nov 08 21:48:08 integro volumio[5391]: info: CoreCommandRouter::executeOnPlugin: volumiodiscovery , saveDeviceInfo 
Nov 08 21:48:08 integro volumio[5391]: info: CoreCommandRouter::volumioPushState 
Nov 08 21:48:08 integro volumio[5391]: info: MRS: Pushing multiroomSync output update for this device 
Nov 08 21:48:08 integro volumio[5391]: info: MRS: Pushing multiroomSync output 
Nov 08 21:48:08 integro volumio[5391]: info: Updating  RAAT Signal Path 
Nov 08 21:48:08 integro volumio[5391]: info: MCU Signalled Playback Inactive 
```

Perhaps it is due to the fact that the ALSA device is opened (though stalled) by camilladsp process?

paolosabatino commented:
Tried also 22khz and mono tracks, both play correctly

balbuze commented:
I will check the wake up with Integro.
Yes, 22kHz play on integro now.
A test to add is on RPI. Set EQ to 15 bands graphic. Play a 22khz file. Last filter (16kHz) should be removed by FusionDsp or CamillaDsp hangs (sample rate must be twice the filter frequency).
Other point, but not sure if related to Spotify or FusionDsp.... Play a 96kHz file (local or Qobuz). Then connect Spotify connect and try to play. Spotify (44.1kHz) play slower because sample rate seems to stay 96kHz.
And both streams are mixed together.

Darmur commented:
> Hmm, I'm having another issue that is related to the MCU on Integro. After turning the unit off, either by idle timeout or by pushing the button, after some seconds (less than a minute) the unit is turned on again.
> 
> `volumio livelog` reports this:
> 
> ```
> Nov 08 21:48:08 integro volumio[5391]: info: MCU Signalled Playback Active 
> Nov 08 21:48:08 integro volumio[5391]: info: MCU Signalled Sleep Mode Disabled 
> Nov 08 21:48:08 integro sudo[7692]:  volumio : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/usr/bin/xset s off -dpms 
> Nov 08 21:48:08 integro sudo[7692]: pam_unix(sudo:session): session opened for user root by (uid=0) 
> Nov 08 21:48:08 integro kernel: fb: osd[0] enable: 1 (Xorg) 
> Nov 08 21:48:08 integro kernel: fb: osd[0] enable: 1 (Xorg) 
> Nov 08 21:48:08 integro sudo[7692]: pam_unix(sudo:session): session closed for user root 
> Nov 08 21:48:08 integro volumio[5391]: info: CoreStateMachine::pushState 
> Nov 08 21:48:08 integro volumio[5391]: info: CorePlayQueue::getTrack 1 
> Nov 08 21:48:08 integro volumio[5391]: info: CoreCommandRouter::executeOnPlugin: volumiodiscovery , saveDeviceInfo 
> Nov 08 21:48:08 integro volumio[5391]: info: CoreCommandRouter::volumioPushState 
> Nov 08 21:48:08 integro volumio[5391]: info: MRS: Pushing multiroomSync output update for this device 
> Nov 08 21:48:08 integro volumio[5391]: info: MRS: Pushing multiroomSync output 
> Nov 08 21:48:08 integro volumio[5391]: info: Updating  RAAT Signal Path 
> Nov 08 21:48:08 integro volumio[5391]: info: MCU Signalled Playback Inactive 
> ```
> 
> Perhaps it is due to the fact that the ALSA device is opened (though stalled) by camilladsp process?

Hi Paolo,

I had the very same issue even without FusionDSP, it was related to a bug with Spotify plugin (v4.0.1). The issue was exactly the same you described, even if Spotify was not used, but installed and activated.
The issue has been solved with the latest stable release (v4.0.3)

Could you please check if it's your case? If yes, please uninstall the old version, reboot, install the new one, reboot again.

Darmur commented:
> I will check the wake up with Integro. Yes, 22kHz play on integro now. A test to add is on RPI. Set EQ to 15 bands graphic. Play a 22khz file. Last filter (16kHz) should be removed by FusionDsp or CamillaDsp hangs (sample rate must be twice the filter frequency). Other point, but not sure if related to Spotify or FusionDsp.... Play a 96kHz file (local or Qobuz). Then connect Spotify connect and try to play. Spotify (44.1kHz) play slower because sample rate seems to stay 96kHz. And both streams are mixed together.

Regarding the issue with Spotify connect after Local/Qobuz at 96kHz... what is the behavior of the system with FusionDSP disabled?

Are you trying to play from Spotify Connect with Local/Qobuz playback still active, or did you stop it before?

paolosabatino commented:
> Then connect Spotify connect and try to play. Spotify (44.1kHz) play slower because sample rate seems to stay 96kHz.

> Regarding the issue with Spotify connect after Local/Qobuz at 96kHz... what is the behavior of the system with FusionDSP disabled?
> Are you trying to play from Spotify Connect with Local/Qobuz playback still active, or did you stop it before?

I had a similar issue even with "stock" FusionDSP (didn't try with no plugins) during the welcome volumio jingle if I start playing a track before the jingle. When the jingle plays and a track is already playing, the two streams interrupt each other and if samplerate is different, one of them has an altered pitch.

paolosabatino commented:
> Could you please check if it's your case? If yes, please uninstall the old version, reboot, install the new one, reboot again.

I will check as soon as possibile, but AFAIR the last time I checked I only had the development FusionDSP plugin installed and enabled. I double checked against the stable FuisionDSP code and wake up was working as expected.

balbuze commented:
tested on RPI4. Switching between different sample rate does not works.
play a local 22kHz
play a Qobuz 96kHz
play a local 22kHz
not played at correct speed.
Camilla must be restarted when sample rate change occurs

paolosabatino commented:
>  weird bis :RPI after a reboot, it works now. Wonder what is the trigger

perhaps it is the node.js cache. Whenever you change index.js - or any other .js - remember to always do `killall node` (or reboot the board) to purge the node.js code cache; otherwise you get a variable rate of weirdness

Darmur commented:
> > Could you please check if it's your case? If yes, please uninstall the old version, reboot, install the new one, reboot again.
> 
> I will check as soon as possibile, but AFAIR the last time I checked I only had the development FusionDSP plugin installed and enabled. I double checked against the stable FuisionDSP code and wake up was working as expected.

OK, thanks for checking

for your information, if the selected source is Streaming and Sleep mode is triggered (orange LED), the device can go to playback mode again (blue LED) for the following reasons:
- The rotary encoders are pressed or turned
- Volumio sends the message SET PLAYBACK 1
- Volumio sends the message SET SLEEP 0
- Volumio sends the message for changing volume
- Volumio sends the message for changing source

paolosabatino commented:
@Darmur 

so apparently volumio is sending a `SET PLAYBACK 1` message even when it is paused and there isn't any playback.

This is a basic session sniffing the serial device:
```
volumio@integro:/dev$ cat ttyS3

<I moved the volume rotary encoder left and right a bit>
PUSH VOLUME 25
PUSH VOLUME 24
PUSH VOLUME 23
PUSH VOLUME 22
PUSH VOLUME 23
PUSH VOLUME 24
PUSH VOLUME 25

<Pause the playback>
PUSH PLAYBACK 0

<Wait for a dozen of seconds without doing anything with encoders and without touching the WebUI>
PUSH PLAYBACK 1
```

So there is a "spourious" playback message that turns the device in playback mode.

Disabling this experimental FusionDSP plugin, the last `PUSH PLAYBACK 1` message is not sent anymore.
I feel quite sure there's a timer that checks against ALSA device status and sends the message which is triggered since camilladsp runs in background and keeps the device open.

paolosabatino commented:
@balbuze 
>  edit again : Spotify seems to be a problem! Especially if play is on when switching from Spotify to local

I did some minimal tests jumping from spotify to local/qobuz tracks but I did not experience yet any sample rate issue: the sampling rate was always changed correctly. The plugin version is 4.0.3 as suggested by @Darmur

Darmur commented:
> @Darmur
> 
> so apparently volumio is sending a `SET PLAYBACK 1` message even when it is paused and there isn't any playback.
> 
> This is a basic session sniffing the serial device:
> 
> ```
> volumio@integro:/dev$ cat ttyS3
> 
> <I moved the volume rotary encoder left and right a bit>
> PUSH VOLUME 25
> PUSH VOLUME 24
> PUSH VOLUME 23
> PUSH VOLUME 22
> PUSH VOLUME 23
> PUSH VOLUME 24
> PUSH VOLUME 25
> 
> <Pause the playback>
> PUSH PLAYBACK 0
> 
> <Wait for a dozen of seconds without doing anything with encoders and without touching the WebUI>
> PUSH PLAYBACK 1
> ```
> 
> So there is a "spourious" playback message that turns the device in playback mode.
> 
> Disabling this experimental FusionDSP plugin, the last `PUSH PLAYBACK 1` message is not sent anymore. I feel quite sure there's a timer that checks against ALSA device status and sends the message which is triggered since camilladsp runs in background and keeps the device open.
> 
> edit: just to add more weirdness...
> 
> 1. play a song
> 2. stop the song
> 3. push the source encoder to bring integro in Sleep mode
> 4. the led becomes orange
> 5. after a few seconds, `PUSH PLAYBACK 1` message is sent by volumio and Integro goes in playback mode and led turns blue
> 6. push the source source to bring integro back in Sleep mode
> 7. There isn't any trace of `PUSH PLAYBACK 1` message anymore and Integro stays in Sleep mode
> 8. side effect: pushing "play" on webUI to restart the track does not trigger `PUSH PLAYBACK 1` message and Integro stays in Sleep mode (but playback time goes on)
> 9. pushing "pause" on webUI stops playback time, integro stays in Sleep mode, no message sent
> 10. pushing "play" on webUI restarts playback and `PUSH PLAYBACK 1` message is sent, Integro wakes up
>     🥲

@volumio any comments?

volumio commented:
I feel quite sure there's a timer that checks against ALSA device status and sends the message which is triggered since camilladsp runs in background and keeps the device open.

Yes, that's the case. But we can ignore this for now and I can fix it easily later on

Darmur commented:
perfect!

@balbuze can you please give a step-by-step guide for reproducing your issue with spotify switching to local playback?

maybe it's something we should fix on spotify side

paolosabatino commented:
Perhaps I found the issue: it is another race condition when the camilladsp process is tore down. In some very rare occurrencies, tearing down takes a bit more than expected. 

Restarting the process before the previous instance is terminated causes a loop in node.js of which I don't really understand the nature yet. I have a couple of ideas for workarounds in my mind and will push a possibile fix shortly.

balbuze commented:
> perfect!
> 
> @balbuze can you please give a step-by-step guide for reproducing your issue with spotify switching to local playback?
> 
> maybe it's something we should fix on spotify side
On integro
If you play a 96kHz local or Qobuz track and without stopping it, connect Spotify and play a track. Spotiify "hide" the other track. Stop Spotify connect, local track still plays but not a the correct speed. Press stop and now play : ok now. So, it seems the issue is on Spotify side.

balbuze commented:
But in overall, it works well! Well done!

balbuze commented:
In conv mode, clipping detection do not works anymore. It returns 3.80dB while stable version returns 14.83dB (correct) with demo files
![image](https://github.com/volumio/volumio-plugins-sources/assets/12850139/b03ea7cd-7c5c-4b8d-bfe9-184634e48b6a)

Darmur commented:
> > perfect!
> > @balbuze can you please give a step-by-step guide for reproducing your issue with spotify switching to local playback?
> > maybe it's something we should fix on spotify side
> 
> On integro If you play a 96kHz local or Qobuz track and without stopping it, connect Spotify and play a track. Spotiify "hide" the other track. Stop Spotify connect, local track still plays but not a the correct speed. Press stop and now play : ok now. So, it seems the issue is on Spotify side.

this is definitely a Spotify problem, already reported from someone with spotify native / spotify connect switch.
@volumio am I wrong?

Darmur commented:
> In conv mode, clipping detection do not works anymore. It returns 3.80dB while stable version returns 14.82dB (correct) with demo files dev version : ![image](https://user-images.githubusercontent.com/12850139/282154149-b03ea7cd-7c5c-4b8d-bfe9-184634e48b6a.png) stable : ![image](https://user-images.githubusercontent.com/12850139/282154405-6ef2a483-17f5-44f9-9c34-3d6b7067b9d5.png)

what if you use the same left and right file?

in the two screenshots they are different

balbuze commented:
in fact this the same files... But the first screenshot is before I press Apply. As this is a set of different sample rate filters, the name displayed is changed to indicate it. But you can try with demo_swap_left file you should get 15.35dB with stable and 3.80dB with dev version

balbuze commented:
This how clipping detection works in FusionDsp when you press Apply.
- selected filter is set
- attenuation is set to 0
- output is redirected to a null device
- A 3 seconds 0dB pink noise is played
- camilladsp.log is read to see clipping messages. The highest value is used.
- An offset ( about 2dB) is added
- the result is set as attenuation
- output is set to normal.
That's it!

paolosabatino commented:
Update:

made another commit with some more corrections trying to make the thing more robust.
I did a quick round of usual tests against qobuz, webradio and local files and did not got any sample rate miss. Only with spotify I get some troubles, but it seems that the plugin is somehow interfering with internal state  trying to play its own content while other content is already playing. Did not investigate further though.

@balbuze I used `fileWriteSync()` in place of `fileWrite()` function [here](https://github.com/volumio/volumio-plugins-sources/pull/274/files/72d315fa8ca62f8ddf24eceec2e7d0a05052e554..a2122fb18321964f50ceefd77c17243780907ed5#diff-5af2730ad779c417b99d391ee235d7ea9a09eb75c7d245a2bfa561308ef46af6R3600-R3606) when writing the camilladsp conf file. That makes the execution flow a faster and more predictable and could also remove the `setTimeout()`.
I wonder if the `readFile()` could be exchanged with `readFileSync()`  in the same function, so perhaps more `setTimeout()`  around could be removed as well (but that would make for another round of changes I guess).

About the clipping detection, perhaps there is the need to adapt the algorithm (I didn't look into) or it is just that camilladsp at the moment is invoked with `debug` flag on instead of `warn` as it was before (see [here](https://github.com/volumio/volumio-plugins-sources/pull/274/files#diff-6944003606b1f7adfeeaa682043abcd04d2758a0c387737656a1e858668e2dc0R14))

balbuze commented:
> About the clipping detection, perhaps there is the need to adapt the algorithm (I didn't look into) or it is just that camilladsp at the moment is invoked with `debug` flag on instead of `warn` as it was before (see [here](https://github.com/volumio/volumio-plugins-sources/pull/274/files#diff-6944003606b1f7adfeeaa682043abcd04d2758a0c387737656a1e858668e2dc0R14))

Yes I do think it is due to timeout... Making it sync would be a good option.

balbuze commented:
Found a specific case. In conv mode, select a filter "kefiq30impulse-x-44100.pcm". FusionFsp switch filters according to samplerate between 44.1 to 96kHz with provided filters->ok
Now if you play a 192kHz track, Camilla returns an error and no sound-> normal behavior
But after that even playing a 44.1kHz track does not work. Play /stop does nothing as Camilla stays in the state. The only way I found is to disable effects in FusionDsp.

paolosabatino commented:
Found a side issue with camilladsp. When the ALSA device is already in use, the process exits but the exit code is success `0`; perhaps this should be notified to camilladsp author:

```
2023-11-12 10:50:06.934447 INFO [src/bin.rs:711] CamillaDSP version 1.0.2
2023-11-12 10:50:06.934637 INFO [src/bin.rs:712] Running on linux, arm
2023-11-12 10:50:06.934860 DEBUG [src/bin.rs:754] Read config file Some("/data/configuration/audio_interface/fusiondsp/camilladsp.yml")
2023-11-12 10:50:06.936484 DEBUG [src/biquad.rs:43] a1=-1.9676843881815393 a2=0.9683681064665667 b0=1 b1=-1.9676843881815393 b2=0.9683681064665667
2023-11-12 10:50:06.936660 DEBUG [src/biquad.rs:43] a1=-1.5979427773292325 a2=0.6257495756964397 b0=1 b1=-1.5979427773292325 b2=0.6257495756964397
2023-11-12 10:50:06.936791 DEBUG [src/biquad.rs:43] a1=-0.9904472828689329 a2=0.39821604860125576 b0=1 b1=-0.9904472828689329 b2=0.39821604860125576
2023-11-12 10:50:06.936919 DEBUG [src/biquad.rs:43] a1=-1.9676843881815393 a2=0.9683681064665667 b0=1 b1=-1.9676843881815393 b2=0.9683681064665667
2023-11-12 10:50:06.937045 DEBUG [src/biquad.rs:43] a1=-1.5979427773292325 a2=0.6257495756964397 b0=1 b1=-1.5979427773292325 b2=0.6257495756964397
2023-11-12 10:50:06.937173 DEBUG [src/biquad.rs:43] a1=-0.9904472828689329 a2=0.39821604860125576 b0=1 b1=-0.9904472828689329 b2=0.39821604860125576
2023-11-12 10:50:06.937284 DEBUG [src/bin.rs:773] Config is valid
2023-11-12 10:50:06.937416 DEBUG [src/socketserver.rs:260] Start websocket server on 127.0.0.1:9876
2023-11-12 10:50:06.937707 DEBUG [src/bin.rs:857] Wait for config
2023-11-12 10:50:06.937840 DEBUG [src/bin.rs:890] Config ready
2023-11-12 10:50:06.938283 DEBUG [src/bin.rs:213] Using channels [true, true]
2023-11-12 10:50:06.938274 ERROR [src/socketserver.rs:290] Failed to start websocket server: Address already in use (os error 98)
2023-11-12 10:50:06.938812 DEBUG [src/bin.rs:333] Capture thread ready to start
2023-11-12 10:50:06.939125 DEBUG [src/filters.rs:450] Build new pipeline
2023-11-12 10:50:06.939336 DEBUG [src/filters.rs:355] Build from config
2023-11-12 10:50:06.939466 DEBUG [src/biquad.rs:43] a1=-1.9676843881815393 a2=0.9683681064665667 b0=1 b1=-1.9676843881815393 b2=0.9683681064665667
2023-11-12 10:50:06.939604 DEBUG [src/biquad.rs:43] a1=-1.5979427773292325 a2=0.6257495756964397 b0=1 b1=-1.5979427773292325 b2=0.6257495756964397
2023-11-12 10:50:06.939730 DEBUG [src/biquad.rs:43] a1=-0.9904472828689329 a2=0.39821604860125576 b0=1 b1=-0.9904472828689329 b2=0.39821604860125576
2023-11-12 10:50:06.939861 DEBUG [src/filters.rs:355] Build from config
2023-11-12 10:50:06.939978 DEBUG [src/biquad.rs:43] a1=-1.9676843881815393 a2=0.9683681064665667 b0=1 b1=-1.9676843881815393 b2=0.9683681064665667
2023-11-12 10:50:06.940197 DEBUG [src/biquad.rs:43] a1=-1.5979427773292325 a2=0.6257495756964397 b0=1 b1=-1.5979427773292325 b2=0.6257495756964397
2023-11-12 10:50:06.940325 DEBUG [src/biquad.rs:43] a1=-0.9904472828689329 a2=0.39821604860125576 b0=1 b1=-0.9904472828689329 b2=0.39821604860125576
2023-11-12 10:50:06.940457 DEBUG [src/processing.rs:19] build filters, waiting to start processing loop
2023-11-12 10:50:06.946562 ERROR [src/bin.rs:344] Playback error: ALSA function 'snd_pcm_open' failed with error 'EBUSY: Device or resource busy'
2023-11-12 10:50:06.946706 DEBUG [src/bin.rs:349] Error while starting, release barrier
2023-11-12 10:50:06.946843 DEBUG [src/bin.rs:352] Wait for capture thread to exit..
2023-11-12 10:50:06.946988 DEBUG [src/filedevice.rs:623] starting captureloop
2023-11-12 10:50:06.947110 DEBUG [src/filedevice.rs:263] starting captureloop
2023-11-12 10:50:06.947480 DEBUG [src/filedevice.rs:299] Exit message received, sending EndOfStream
2023-11-12 10:50:06.947770 DEBUG [src/bin.rs:915] Restarting with new config
2023-11-12 10:50:06.947900 DEBUG [src/bin.rs:857] Wait for config
2023-11-12 10:50:06.948006 DEBUG [src/bin.rs:860] No config and not in wait mode, exiting!
volumio@integro:/tmp$ echo $?
0
```

paolosabatino commented:
Update:

I should have fixed the missed resume after the convolution filter is applied to 192khz tracks. camilladsp does not start because of the missing convolution filter file and this also caused the "loop of process respawning due to unknown nature". That caused the FIFO to stall (which, in turn, makes impossible to restart playing because the whole alsa pipeline stalls) and also triggered the loop of process respawns.

I'm not very fond of the [solution to solve the fifo stall](https://github.com/volumio/volumio-plugins-sources/pull/274/files#diff-6944003606b1f7adfeeaa682043abcd04d2758a0c387737656a1e858668e2dc0R97) as it may be source of some other kind of issues, although during my tests nothing bad happened yet.

Both issues should be fixed right now.

balbuze commented:
Awesome! I try tonight!
You're genius 😜

balbuze commented:
Found a specific case. In conv mode, select a filter "kefiq30impulse-x-44100.pcm". FusionFsp switch filters according to samplerate between 44.1 to 96kHz with provided filters->ok
Now if you play a 192kHz track, Camilla returns an error and no sound-> normal behavior
But after that even playing a 44.1kHz track does not work. Play /stop does nothing as Camilla stays in the state. The only way I found is to disable effects in FusionDsp. 

> I should have fixed the missed resume after the convolution filter is applied to 192khz tracks

Perfect! No more blocking issue! Very very nice! Can we get a toastmessage in this case : "Error, no filter for this sample rate" or something like that

Darmur commented:
awesome work, thanks to you both!

@balbuze do you think is good enough for merging the pull-request, and create a new stable version of the plugin?

this would allow us to create new images with updated mpd

balbuze commented:
Nearly ready... Remains clipping detection in conv mode (see above). Once this fixed...
But in overall really good job! I hope it works well with updated mpd ;-)

paolosabatino commented:
@balbuze it looks like the clipping detection works.
I changed camilla logging level from `debug` to `warn`  (`warn` is the current logging level on stable fusiondsp code) and it looks like it is working as expected:

![immagine](https://github.com/volumio/volumio-plugins-sources/assets/15271078/e4fe2e14-50c4-492d-b748-dc0b33dafb37)

The only caveat, which I don't know if it is present in stable fusiondsp, is that you must first stop playing the current tracker, then apply the convolution filter, otherwise you get wrong numbers (I get 3.80db attenuation, for example). 

Perhaps it makes sense to stop playing first when the user pushes "apply" in case conv filter is selected?

balbuze commented:
Yes, playback is pauses.
So super!
Ready for a release then?😜

paolosabatino commented:
Apparently everything is set; is there the need for a version bump?

balbuze commented:
I would update to 1.0.50.
For me version 2.x.x comes with the future CamillaDsp 2.x.
Need to update Readme.md to add changelog. 
And advertising on community thread.
I can do it if you want.
But not before tonight

paolosabatino commented:
No problems in updating the version number, I will set 1.0.50 as soon as I can.
Then it should be ready. I also did a quick test with current MPD 0.21 and it worked without apparent glitches.

paolosabatino commented:
done, version has been bumped to 1.0.50

Darmur commented:
@balbuze I still see 1.0.49 as latest available version on the plugin store, could you please publish the new 1.0.50?

balbuze commented:
I'm on it... Please wait few minutes ;-)

balbuze commented:
Huumm a bad replace cmd... Wait 1.0.51....

balbuze commented:
1.0.51 submitted as beta. Please check if ok as I had some stupid problems ... ;-)

paolosabatino commented:
> 1.0.51 submitted as beta. Please check if ok as I had some stupid problems ... ;-)

Checked, I first installed the stable 1.0.49 version, then removed and installed 1.0.51 beta version.
At first it was not working, perhaps node.js still kept the previous version in cache. Had to reboot Integro and finally it worked as expected.

Darmur commented:
yes, I don't know why but after updating a plugin a reboot it's mandatory to have the new code loaded properly

balbuze commented:
In beta mode only I think

paolosabatino commented:
Hello, I was looking at this post on the forum: https://community.volumio.com/t/fusiondsp-the-complete-dsp-center-for-volumio3/50474/1093 and I got the feel the [stall solution](https://github.com/volumio/volumio-plugins-sources/pull/274/files#diff-6944003606b1f7adfeeaa682043abcd04d2758a0c387737656a1e858668e2dc0R97) maybe has something to do with it; although it looks more reasonable to me if the stolen samples are less than 1 seconds rather than 10 seconds, which is way too much IMHO.

Perhaps it happens only on RPi4 because of the much faster processor than Integro; perhaps mpd is filling buffers much earlier and faster than on the vim3l

balbuze commented:
Yes, maybe on rpi4... On Integro I didn't notice that.
I have to test on primo and x86 too.

paolosabatino commented:
@balbuze thanks, it would be very useful. I will think about a possible alternative solution to the stall workaround. Perhaps just adding a `count=1` to the `dd` commandline is sufficient to avoid stealing too much samples, and perhaps reducing the buffersize to a mere `bs=1k` would make it non-noticeable at all

paolosabatino commented:
@balbuze I made a draft PR with a possible fix: https://github.com/volumio/volumio-plugins-sources/pull/276 - in pratice the FIFO is emptied only when the camilladsp process terminated with an error, which is the case the convolution filter file is not available.
